src/garage/torch/algos/td3.py

```diff
@@ -0,0 +1,429 @@
+            -1, 1)
+        actions = samples_data['actions'].to(global_device())
+        observations = samples_data['observations'].to(global_device())
+        next_observations = samples_data['next_observations'].to(
+            global_device())
+
+        next_inputs = next_observations
+        inputs = observations
+        with torch.no_grad():
+            # Select action according to policy and add clipped noise
+            noise = (torch.randn_like(actions) * self._policy_noise).clamp(
+                -self._policy_noise_clip, self._policy_noise_clip)
+            next_actions = (self._target_policy(next_inputs) + noise).clamp(
+                -self._max_action, self._max_action)
+
+            # Compute the target Q value
+            target_Q1 = self._target_qf_1(next_inputs, next_actions)
+            target_Q2 = self._target_qf_2(next_inputs, next_actions)
+            target_q = torch.min(target_Q1, target_Q2)
+            target_Q = rewards * self._reward_scaling + (
+                1. - terminals) * self._discount * target_q
+
+        # Get current Q values
+        current_Q1 = self._qf_1(inputs, actions)
+        current_Q2 = self._qf_2(inputs, actions)
+        current_Q = torch.min(current_Q1, current_Q2)
+
+        # Compute critic loss
+        critic_loss = F.mse_loss(current_Q1, target_Q) + F.mse_loss(
+            current_Q2, target_Q)
+
+        # Optimize critic
+        self._qf_optimizer_1.zero_grad()
+        self._qf_optimizer_2.zero_grad()
+        critic_loss.backward()
+        self._qf_optimizer_1.step()
+        self._qf_optimizer_2.step()
+
+        # Deplay policy updates
+        if grad_step_timer % self._update_actor_interval == 0:
+            # Compute actor loss
+            actions = self.policy(inputs)
+            self._actor_loss = -self._qf_1(inputs, actions).mean()
+
+            # Optimize actor
+            self._policy_optimizer.zero_grad()
+            self._actor_loss.backward()
+            self._policy_optimizer.step()
+
+            # update target networks
+            self._update_network_parameters()
+
+        return (critic_loss.detach(), target_Q, current_Q.detach(),
+                self._actor_loss.detach())
+
+    def _evaluate_policy(self):
+        """Evaluate the performance of the policy via deterministic rollouts.
+
+        Statistics such as (average) discounted return and success rate are
+            recorded.
+
+        Returns:
+            TrajectoryBatch: Evaluation trajectories, representing the best
+                current performance of the algorithm.
+
+        """
+        return obtain_evaluation_episodes(
+            self.exploration_policy,
+            self._eval_env,
+            num_eps=self._num_evaluation_episodes,
+            deterministic=False)
+
+    def _update_network_parameters(self):
+        """Update parameters in actor network and critic networks."""
+        for target_param, param in zip(self._target_qf_1.parameters(),
```

Comment:
I think I've seen this same loop in multiple algorithms now. Can you pull it out into a function like `garage.torch.soft_update_model(target_model, source_model, tau)`? You don't need to refactor the other algorithms.